llvm/lib/Target/NVPTX/NVPTXProxyRegErasure.cpp

```diff
@@ -78,7 +78,11 @@ bool NVPTXProxyRegErasure::runOnMachineFunction(MachineFunction &MF) {
         assert(InOp.isReg() && "ProxyReg input should be a register.");
         assert(OutOp.isReg() && "ProxyReg output should be a register.");
         RemoveList.push_back(&MI);
-        RAUWBatch.try_emplace(OutOp.getReg(), InOp.getReg());
+        Register replacement = InOp.getReg();
+        // Check if the replacement itself has been replaced.
+        if (auto it = RAUWBatch.find(replacement); it != RAUWBatch.end())
+          replacement = it->second;
```

Comment:
Will it deal with longer chains of ProxyReg? It would be great to add a test for that.

Perhaps we should iteratively apply replacements until they produce no more changes.

Reply:
This will deal with longer chains, because it feeds forward the leaf Register. I'll add a test.

```
%0 = proxy %src // %0 mapped to %src
%1 = proxy %0 // since %0 is mapped to %src, %1 is mapped to %src
%2 = proxy %1 // since %1 is mapped to %src, %2 is mapped to %src
```

Reply:
Added